File: src/backend/src/main/java/com/myus/service/FaqService.java
package com.myus.service;

import com.myus.dto.FaqResponse;

import org.springframework.data.domain.Page;

import java.util.List;

/**
 * Service contract for the centralized FAQ library (UC10).
 *
 * <p>Provides searchable, categorized access to published FAQ content covering
 * university policies, academic rules, and IT support, plus lightweight
 * feedback collection ("Helpful" / "Not Helpful") on individual answers.</p>
 */
public interface FaqService {

    /**
     * Distinct categories present among published FAQ articles, sorted alphabetically.
     */
    List<String> getCategories();

    /**
     * Search/browse published FAQ articles with keyword and category filtering.
     *
     * <p>The keyword search tolerates typos (via edit-distance matching) and
     * recognizes common keyword synonyms (e.g. "drop a class" ≈ "withdraw
     * from course"), so results stay relevant even with imprecise queries.</p>
     *
     * @param search   optional free-text keyword search
     * @param category optional category filter
     * @param page     zero-based page index
     * @param size     number of items per page
     * @return paginated, relevance-ranked FAQ summaries
     */
    Page<FaqResponse> search(String search, String category, int page, int size);

    /**
     * Most helpful published FAQ articles, used to suggest popular topics
     * when a search returns no results (AF1).
     */
    List<FaqResponse> getPopular(int limit);

    /**
     * Full detail for a single published FAQ article, including related
     * questions from the same category (AF4).
     *
     * @throws com.myus.exception.ResourceNotFoundException if not found or unpublished
     */
    FaqResponse getById(Long faqId);

    /**
     * Record a "Helpful" / "Not Helpful" rating for an answer (AF3).
     *
     * @return the updated helpful/not-helpful counts
     */
    FaqResponse submitFeedback(Long faqId, boolean helpful);
}

File: src/frontend/src/services/faqService.ts
import axiosInstance from '../api/axiosInstance';
import type { FAQArticle, PagedResponse } from '../types';

// ============================================================
// FAQ Service – wraps /api/faq (UC10 centralized FAQ library)
// ============================================================

/** GET /api/faq/categories – distinct FAQ categories. */
export async function getFaqCategories(): Promise<string[]> {
  const { data } = await axiosInstance.get<string[]>('/api/faq/categories');
  return data;
}

/** GET /api/faq – search/browse FAQ entries with optional keyword & category filters. */
export async function searchFaqs(params?: {
  search?: string;
  category?: string;
  page?: number;
  size?: number;
}): Promise<PagedResponse<FAQArticle>> {
  const { data } = await axiosInstance.get<PagedResponse<FAQArticle>>('/api/faq', { params });
  return data;
}

/** GET /api/faq/popular – most helpful FAQ entries (suggested when search yields no results). */
export async function getPopularFaqs(limit: number = 5): Promise<FAQArticle[]> {
  const { data } = await axiosInstance.get<FAQArticle[]>('/api/faq/popular', { params: { limit } });
  return data;
}

/** GET /api/faq/{id} – full answer detail, including related questions. */
export async function getFaqById(faqId: string | number): Promise<FAQArticle> {
  const { data } = await axiosInstance.get<FAQArticle>(`/api/faq/${faqId}`);
  return data;
}

/** POST /api/faq/{id}/feedback – rate an answer "Helpful" / "Not Helpful". */
export async function submitFaqFeedback(
  faqId: string | number,
  helpful: boolean
): Promise<FAQArticle> {
  const { data } = await axiosInstance.post<FAQArticle>(`/api/faq/${faqId}/feedback`, { helpful });
  return data;
}
